src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1123,19 +1123,25 @@ internal ValueTask<IBackendMessage> ReadMessage(bool async, DataRowLoadingMode d
                             Debug.Assert(msg != null, "Message is null for code: " + messageCode);
                             return msg;
                         }
-                        catch (NpgsqlException e) when (!readingNotifications2 && e.InnerException is TimeoutException && originalTimeoutException is null)
+                        catch (NpgsqlException e) when (!readingNotifications2 && e.InnerException is TimeoutException)
                         {
-                            // We have got a timeout while not reading the async notifications - trying to cancel a query
-                            try
+                            if (originalTimeoutException is null)
```

Comment:
nit: I'd invert this check, this way the bigger block of code doesn't have to be indented.